### PR TITLE
🔖example: adapt

### PR DESCRIPTION
**Pull request by Abhrant**
This PR adds the examples for Adapt.

### CODE REVIEW

**Comment by shwu-nyunai** [comment]
any reasons, we have the examples using dockers? and if docker, why not with the cli directly?

File: examples/adapt/image_segmentation/README.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+$ docker run -it -d --gpus all -v /dev/shm:/dev/shm -v $(pwd):/workspace --name nyuntam_adapt_docker --network=host nyunadmin/nyunzero_adapt:v0.1 bash 
+
+$ docker exec -it nyuntam_adapt_docker /bin/bash
+
+$ cd /workspace/nyuntam
+```
+
+## Dataset 
+
+For this experiment we will be using the [CityScapes](https://www.kaggle.com/datasets/ipythonx/stanford-background-dataset) dataset. In this dataset we will only be using the "regions" as the segmentation maps. The data folder structure is formatted in the folowing way:
```
Comment (shwu-nyunai) [inline examples/adapt/image_segmentation/README.md:54]: Do we have any formatting/modifications on this dataset to account only for the "regions" as segmentation? If yes, we can add details on the formatting step(s) too.

File: examples/adapt/image_segmentation/README.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+BEGIN: 0
+END: 50
+
+#MMSEG SPECIFIC ARGUMENTS
+amp :  False
+resume  : #False # ['auto']
+auto_scale_lr :  False
+cfg_options  : None
+launcher :  none
+dest_root : './.mmdet_cache'   #This saves the checkpoints and .py files for the model congigs and mmdet logs
```
Comment (shwu-nyunai) [inline examples/adapt/image_segmentation/README.md:119]: nits -
spelling mistake in the comment. `model congigs -> model configs`

File: examples/adapt/image_segmentation/README.md
Comment (shwu-nyunai) [inline examples/adapt/image_segmentation/README.md:1]: Do we also have any evaluation or downstream steps after fine-tuning?

File: examples/adapt/image_segmentation/config.yaml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+BEGIN: 0
+END: 50
+
+#MMSEG SPECIFIC ARGUMENTS
+amp :  False
+resume  : #False # ['auto']
+auto_scale_lr :  False
+cfg_options  : None
+launcher :  none
+dest_root : './.mmdet_cache'   #This saves the checkpoints and .py files for the model congigs and mmdet logs
```
Comment (shwu-nyunai) [inline examples/adapt/image_segmentation/config.yaml:41]: nits.
same as above.

File: examples/adapt/object_detection/README.md
Comment (shwu-nyunai) [inline examples/adapt/object_detection/README.md:1]: 1. Same as above, can this follow any evaluation or other downstream steps after ft? Or is it not needed? 
2. Should we be adding a comparison in the training time with/without DDP?

cc: @Arnav0400

File: examples/adapt/object_detection/README.md [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+ - [Installation](#installation)
+ - [Dataset](#dataset)
+ - [Configuration](#configuration)
+ - [Adapting the model](#adapting-the-model)
+ - [Conclusion](#conclusion)
+ 
+
+## Introduction 
+
+In this example, we'll be applying the LoRA technique to fine-tune the RTMDet model on the Face-Det dataset. The training process will be distributed across 2 GPUs using Distributed Data Parallel (DDP) to maximize efficiency. LoRA's ability to introduce low-rank adaptations allows for targeted model updates, significantly reducing memory and computational requirements while increasing OOD generalization performance.
```
Comment (shwu-nyunai) [inline examples/adapt/object_detection/README.md:19]: Can expand OOD.

```suggestion
In this example, we'll be applying the LoRA technique to fine-tune the RTMDet model on the Face-Det dataset. The training process will be distributed across 2 GPUs using Distributed Data Parallel (DDP) to maximize efficiency. LoRA's ability to introduce low-rank adaptations allows for targeted model updates, significantly reducing memory and computational requirements while increasing Out-of-Distribution (OOD) generalization performance. 
```

File: examples/adapt/object_detection/README.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+│   └── Adapt
+│       └── OBJDET
+├── logs
+    └── Adapt
+        └── OBJDET
+            └── log.log
+
+```
+
+LoRa decreases the toal trainable parameters of the model by freezing the original model weights and updating LoRA adapters only.
```
Comment (shwu-nyunai) [inline examples/adapt/object_detection/README.md:168]: nits -
spelling mistake "total".
Lets do a sanity check on spelling mistakes and grammars across the examples.

File: examples/adapt/object_detection/README.md [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+```
+
+LoRa decreases the toal trainable parameters of the model by freezing the original model weights and updating LoRA adapters only. 
+```log 
+08/05/2024 23-04-20 - INFO - nyuntam_adapt.core.base_algorithm - trainable params: 1491594 || all params: 6009438 || trainable%: 24.82
+```
+
+
+This is a sample log the experiment logs  :
```
Comment (shwu-nyunai) [inline examples/adapt/object_detection/README.md:174]:
```suggestion
This is a sample of the experiment logs  : 
```

File: examples/adapt/text_generation/README.md
Comment (shwu-nyunai) [inline examples/adapt/text_generation/README.md:1]: can we have an eval for this? some task that the dataset in concern increases the performance for. or not needed?

File: examples/adapt/text_generation/README.md [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+user_data/
+├── jobs
+│   └── Adapt
+│       └── SUMM
+│           └── merged_model_state_dict.pth
+├── logs
+    └── Adapt
+        └── SUMM
+            └── log.log
+
```
Comment (shwu-nyunai) [inline examples/adapt/text_generation/README.md:200]: is the tree correct for this algorithm? I see SUMM as the dir name for logs and o/p files.

**Review by shwu-nyunai — CHANGES_REQUESTED** [review]
Overall,
- do we need downstream tasks for the adaption algorithms? like, perf evals for finetuning and train time comparison if using dist training.
- any specific reasons why docker is the default flow? we can instead use the cli

**Review by shwu-nyunai — CHANGES_REQUESTED** [review]
also, what about the other comments? 
@Abhrant

File: examples/adapt/text_generation/README.md [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+├── jobs
+│   └── Adapt
+│       └── 100
+├── logs
+    └── Adapt
+        └── 100
+            └── log.log
+
+```
+The output model will be stored in `user_data/jobs/Adapt/SUMM/` directory and the final directory structure will be:
```
[the remark this answers is not in the document]
Comment (Abhrant) [reply examples/adapt/text_generation/README.md]: done

**Comment by Abhrant** [comment]
@shwu-nyunai , This workflow currently has the most suitable workflowfor adapt (the docker workflow) which can slo be done via the nyuntam repo. We are not showing the cli workflow here as far as I saw in other examples as well. So, These examples show the use of the nyuntam repo (and the way to setup the env is the docker)

**Comment by shwu-nyunai** [comment]
> @shwu-nyunai , This workflow currently has the most suitable workflowfor adapt (the docker workflow) which can slo be done via the nyuntam repo. We are not showing the cli workflow here as far as I saw in other examples as well. So, These examples show the use of the nyuntam repo (and the way to setup the env is the docker)

What I meant is we can either do a docker flow example (which must use the nyuncli) or local installation; 
Vision examples uses docker and hence the example line out Nyun CLI flow. text generation lines out a local installation hence not docker dependent. Adapt should also either run with the cli or the local installation step right? Why the docker if cli does the same thing, in fact does it without any hassle of starting or maintaining the containers; 

cc: @Arnav0400 what is your opinion?

**Comment by Abhrant** [comment]
@shwu-nyunai , The user doesn't know what the cli uses. These example shows how to use the nyuntam repo. The docker is just used to setup the environment and has no code. We can have a env workflow liek in text_gen but I think the docker is also fine and that is the most tested for now. So, I have kept that. 

cc @Arnav0400

**Comment by shwu-nyunai** [comment]
closing in lieu #10

**Pull request closed without merging**